### PR TITLE
Don't modify header for CONNECT style proxies.

### PR DESCRIPTION
**Pull request by psschroeter**
Nginx at least does not like it in certain cases, such as when basic
auth is used to create a session token.

### CODE REVIEW

**Comment by psschroeter** [comment]
Yes more nil port bugs! Well I fixed the socks one also. I see its sort of addressed in some other pull request as well. So if you're using addressable::uri, it really seems you should always be using .inferred_port in most places. .port is better than .inferred_port almost never.

You mentioned tests in that PR, but I'm not sure what you meant? Don't see any many unit style tests to test something like a function returns a certain value.  Not sure if I see an easy way to add in an integration type test for this beyond what's there as it doesn't seem like the the stallion spec server speaks https?

File: lib/em-http/http_encoding.rb [hunk 1]
```diff
@@ -51,7 +51,10 @@ def encode_request(method, uri, query, proxy)
 
       # Non CONNECT proxies require that you provide the full request
       # uri in request header, as opposed to a relative path.
-      query = uri.join(query) if proxy
+      # Don't modify the header with CONNECT proxies. It's unneeded and will
+      # cause 400 Bad Request errors with many standard setups.
+      is_connect = proxy && (proxy[:type] == :http || proxy[:type].nil?) && uri.scheme == "https"
+      query = uri.join(query) if proxy && !is_connect
```
Comment (igrigorik) [inline lib/em-http/http_encoding.rb]: I'm a bit dubious of this one.. as it seems we'd fix one thing, break another? If I remember correctly, Apache/Squid require that you send the full request URL in connect, which is why this logic is in place.. but nginx errors out? 

To make things even more interesting, the spec does seem to indicate that we should only send the hostname and port number: http://tools.ietf.org/html/rfc7231#section-4.3.6

That aside.. In terms of implementation, I think we can make this cleaner: pass connopts object to this method and check for "http_proxy?" instead of duplicating logic.

File: lib/em-http/http_connection.rb [hunk 2]
```diff
@@ -152,9 +152,9 @@ def connection_completed
       @peer = @conn.get_peername
 
       if @connopts.socks_proxy?
-        socksify(client.req.uri.host, client.req.uri.port, *@connopts.proxy[:authorization]) { start }
+        socksify(client.req.uri.host, client.req.uri.inferred_port, *@connopts.proxy[:authorization]) { start }
       elsif @connopts.connect_proxy?
-        connectify(client.req.uri.host, client.req.uri.port, *@connopts.proxy[:authorization]) { start }
+        connectify(client.req.uri.host, client.req.uri.inferred_port, *@connopts.proxy[:authorization]) { start }
```
Comment (igrigorik) [inline lib/em-http/http_connection.rb]: Can we split this into a separate commit? This appears to be a dupe of #278.

File: lib/em-http/http_encoding.rb (hunk 1 above)
Comment (psschroeter) [inline lib/em-http/http_encoding.rb]: Sure, I like passing in connopts and will do that. I guess I wasn't sure if I permission to change up any internal interfaces. 

Not sure if your memory is correct on the apache/squid thing. For Non-connect HTTP, the full url needs to be in the request for sure since it needs to know where to setup the connection to and has no where else to get it other than parsing the http message headers. 

For connect, two requests are sent, the first "CONNECT host:port HTTP/version" message to set up the connection then the actual http request.  The header for the first message isn't  affected by this code other than fixing the nil port thing. So I don't see it being possible for this code to break apache/squid based proxies. The code change affects the header of the second message, which I don't think it should. In the connect scenario the proxy is acting as a dumb tunnel for ssl data so there's no need to rewrite headers of requests passing though.

File: lib/em-http/http_connection.rb (hunk 2 above)
Comment (psschroeter) [inline lib/em-http/http_connection.rb]: Alright split. Using connopts now above. Should be ready now.

**Comment by igrigorik** [comment]
:+1: ... and thanks for your patience :)
